### ISSUE TITLE
PANIC: unprotected error in call to Lua API (not enough memory)

### ISSUE THREAD

**Issue by atav32** [billzorn/mtg-rnn#1]
Not sure if this is even a bug with the repo, since it seems like a Lua error. Just wondering if anyone's run into this issue before.

#### Command
`th train.lua -rnn_size 256 -num_layers 3 -seq_length 200 -data_dir data/mtgencode-std -checkpoint_dir cv/mtgencode-std/ -eval_val_every 1000 -seed 7767`

#### Result
``` Lua
...
997/12000 (epoch 4.154), train_loss = 0.55129399, grad/param norm = 1.9592e-01, time/batch = 2.6165s
998/12000 (epoch 4.158), train_loss = 0.55412990, grad/param norm = 2.0948e-01, time/batch = 2.6173s
999/12000 (epoch 4.163), train_loss = 0.55656064, grad/param norm = 2.2091e-01, time/batch = 2.5914s
evaluating loss over split index 2
1/13...
PANIC: unprotected error in call to Lua API (not enough memory)
```

-----

- OS X 10.11.6 El Capitan 
  - (*I know, I know, the repo says to use Linux, but everything seems to run fine*)
- Memory: 16GB 1600 MHz DDR3
- Graphics: Intel Iris Pro 1536 MB

**Comment by rmmilewi** [comment]
I take it the computation is all on CPU and not GPU? I was about to say that it's possible to exhaust the memory bandwidth of low-end graphics cards very quickly, but that doesn't seem to be the problem here.

I recall that LuaJIT has a memory limit of its own, though there are workarounds for that. That might be what's happening. If you have 200 timesteps, and you're unfolding the [entire graph over those time steps like this](http://d3kbpzbmcynnmx.cloudfront.net/wp-content/uploads/2015/09/rnn.jpg), then yes, there's a lot of memory being used there. 

On the other hand, we should be garbage collecting frequently enough such that it isn't a problem (or are we?), so I'm rather surprised to see the error.

**Comment by atav32** [comment]
@rmmilewi I believe the computation was done on GPU 0 since I left out `-gpuid -1`. 

I was also thinking that maybe the code didn't garbage collect often enough, so I set `-eval_val_every` to `200` instead of `1000` (which was suggested by the [mtgencode tutorial](https://github.com/billzorn/mtgencode#training-a-neural-net)) and it avoids the problem 😂

**Comment by atav32** [comment]
also this ticket might help https://github.com/karpathy/char-rnn/issues/80
